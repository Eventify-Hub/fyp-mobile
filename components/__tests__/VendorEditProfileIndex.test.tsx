import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import EditProfileScreen from "../vendoreditprofile/VendorEditProfileIndex"; // adjust path
import * as store from "@/store";
import patchUpdateProfile from "@/services/patchUpdateProfile";
import { useRouter } from "expo-router";
import renderer from "react-test-renderer"; // at the top of your file

jest.mock("@/store");
jest.mock("@/services/patchUpdateProfile");
jest.mock("expo-router", () => ({
  useRouter: jest.fn(),
}));

describe("EditProfileScreen", () => {
  const mockPush = jest.fn();
  const mockBack = jest.fn();

 beforeEach(() => {
   (useRouter as jest.Mock).mockReturnValue({
     push: mockPush,
     back: mockBack,
   });

   (store.getSecureData as jest.Mock).mockResolvedValue(
     JSON.stringify({
       name: "John Doe",
       email: "john@example.com",
       phoneNumber: "3001234567",
       address: "123 Street",
       userId: "user123",
     })
   );

   // ✅ Add this to prevent alert crash
   global.alert = jest.fn();
 });

  it("renders correctly and shows user initials in avatar", async () => {
    const { getByText, findByText } = render(<EditProfileScreen />);
    expect(await findByText("Edit Profile")).toBeTruthy();
    expect(await findByText("J")).toBeTruthy(); // Avatar letter
  });

  it("loads and displays user data in input fields", async () => {
    const { findByDisplayValue, getByTestId } = render(<EditProfileScreen />);

    expect(await findByDisplayValue("John Doe")).toBeTruthy();
    expect(await findByDisplayValue("john@example.com")).toBeTruthy();
    expect(await findByDisplayValue("3001234567")).toBeTruthy();

    // Use testID to avoid conflict with duplicate values
    expect(getByTestId("input-address").props.value).toBe("123 Street");
  });


  it("updates name field on typing", async () => {
    const { getByPlaceholderText } = render(<EditProfileScreen />);
    const nameInput = getByPlaceholderText("Enter your name");
    fireEvent.changeText(nameInput, "Jane Doe");
    expect(nameInput.props.value).toBe("Jane Doe");
  });

  it("selects staff option", () => {
    const { getByText } = render(<EditProfileScreen />);
    const maleButton = getByText("MALE");
    fireEvent.press(maleButton);
    expect(maleButton.props.style).toEqual(
      expect.arrayContaining([expect.objectContaining({ color: "#FFF" })])
    );
  });

  it("selects refund policy", () => {
    const { getByText } = render(<EditProfileScreen />);
    const policyOption = getByText("REFUNDABLE");
    fireEvent.press(policyOption);
    expect(policyOption.props.style).toEqual(
      expect.arrayContaining([expect.objectContaining({ color: "#FFF" })])
    );
  });
 
  it("calls patchUpdateProfile and saves on SAVE press", async () => {
    (patchUpdateProfile as jest.Mock).mockResolvedValue({
      name: "John Updated",
      email: "john@example.com",
      phoneNumber: "3001234567",
      address: "New Address",
      userId: "user123",
    });

    const { getByText, getByPlaceholderText, getByTestId } = render(
      <EditProfileScreen />
    );

    fireEvent.changeText(
      getByPlaceholderText("Enter your name"),
      "John Updated"
    );

    // Avoid ambiguous placeholder — use testID
    fireEvent.changeText(getByTestId("input-address"), "New Address");

    fireEvent.press(getByText("SAVE"));

    await waitFor(() => {
      expect(patchUpdateProfile).toHaveBeenCalledWith(
        "user123",
        expect.objectContaining({
          name: "John Updated",
          address: "New Address",
        })
      );
    });
  });

  it("navigates back when back button is pressed", () => {
    const { getByText } = render(<EditProfileScreen />);
    fireEvent.press(getByText("< Back"));
    expect(mockBack).toHaveBeenCalled();
  });

    it("has no staff selected initially", () => {
      const { getByText } = render(<EditProfileScreen />);
      const maleButton = getByText("MALE");
      expect(maleButton.props.style).not.toEqual(
        expect.arrayContaining([expect.objectContaining({ color: "#FFF" })])
      );
    });

    it("changes refund policy selection correctly", () => {
      const { getByText } = render(<EditProfileScreen />);
      const refundable = getByText("REFUNDABLE");
      const nonRefundable = getByText("NON-REFUNDABLE");

      fireEvent.press(refundable);
      fireEvent.press(nonRefundable);

      expect(nonRefundable.props.style).toEqual(
        expect.arrayContaining([expect.objectContaining({ color: "#FFF" })])
      );
      expect(refundable.props.style).not.toEqual(
        expect.arrayContaining([expect.objectContaining({ color: "#FFF" })])
      );
    });

    it("allows only one staff selection at a time", () => {
      const { getByText } = render(<EditProfileScreen />);
      const male = getByText("MALE");
      const female = getByText("FEMALE");

      fireEvent.press(male);
      expect(male.props.style).toEqual(
        expect.arrayContaining([expect.objectContaining({ color: "#FFF" })])
      );

      fireEvent.press(female);
      expect(female.props.style).toEqual(
        expect.arrayContaining([expect.objectContaining({ color: "#FFF" })])
      );
      expect(male.props.style).not.toEqual(
        expect.arrayContaining([expect.objectContaining({ color: "#FFF" })])
      );
    });

    it("updates description and cities fields when address is changed", () => {
      const { getByTestId } = render(<EditProfileScreen />);
      const addressInput = getByTestId("input-address");
      const descriptionInput = getByTestId("input-description");
      const citiesInput = getByTestId("input-cities");

      fireEvent.changeText(addressInput, "New Shared Address");

      expect(descriptionInput.props.value).toBe("New Shared Address");
      expect(citiesInput.props.value).toBe("New Shared Address");
    });

    it("saves updated data after changing all form fields", async () => {
      (patchUpdateProfile as jest.Mock).mockResolvedValue({
        name: "Ali Vendor",
        email: "ali@vendor.com",
        phoneNumber: "3011122233",
        address: "New Zone",
        userId: "user123",
      });

      const { getByText, getByPlaceholderText, getByTestId } = render(
        <EditProfileScreen />
      );

      fireEvent.changeText(
        getByPlaceholderText("Enter your name"),
        "Ali Vendor"
      );
      fireEvent.changeText(
        getByPlaceholderText("Enter your email"),
        "ali@vendor.com"
      );
      fireEvent.changeText(getByTestId("input-address"), "New Zone");
      fireEvent.changeText(
        getByPlaceholderText("Enter phone number"),
        "3011122233"
      );

      fireEvent.press(getByText("SAVE"));

      await waitFor(() => {
        expect(patchUpdateProfile).toHaveBeenCalledWith(
          "user123",
          expect.objectContaining({
            name: "Ali Vendor",
            email: "ali@vendor.com",
            phoneNumber: "3011122233",
            address: "New Zone",
          })
        );
      });
    });

    it("shows correct initial in avatar based on loaded name", async () => {
      const { findByText } = render(<EditProfileScreen />);
      expect(await findByText("J")).toBeTruthy(); // From "John Doe"
    });

    it("shows alert if patchUpdateProfile fails", async () => {
      (patchUpdateProfile as jest.Mock).mockRejectedValue(
        new Error("Network Error")
      );

      const { getByText, getByTestId, getByPlaceholderText } = render(
        <EditProfileScreen />
      );
      fireEvent.changeText(
        getByPlaceholderText("Enter your name"),
        "Test User"
      );
      fireEvent.changeText(getByTestId("input-address"), "Somewhere");

      fireEvent.press(getByText("SAVE"));

      await waitFor(() => {
        expect(global.alert).toHaveBeenCalledWith(
          "Failed to save profile. Please try again."
        );
      });
    });

it("renders all refund policy options", () => {
  const { getByText } = render(<EditProfileScreen />);
  expect(getByText("REFUNDABLE")).toBeTruthy();
  expect(getByText("NON-REFUNDABLE")).toBeTruthy();
  expect(getByText("PARTIALLY REFUNDABLE")).toBeTruthy();
});

it("renders all staff options", () => {
  const { getByText } = render(<EditProfileScreen />);
  expect(getByText("MALE")).toBeTruthy();
  expect(getByText("FEMALE")).toBeTruthy();
  expect(getByText("TRANSGENDER")).toBeTruthy();
});

it("shows error or keeps phone field unchanged when alphabets are entered", () => {
  const { getByPlaceholderText } = render(<EditProfileScreen />);
  const phoneInput = getByPlaceholderText("Enter phone number");

  fireEvent.changeText(phoneInput, "abc123");

  // Allowing user to type, but expecting your logic to prevent saving or to show validation later.
  expect(phoneInput.props.value).toBe("abc123"); // ✅ Correct expectation

  // Add this if you eventually show an error
  // const { getByText } = render(<EditProfileScreen />);
  // expect(getByText("Invalid phone number")).toBeTruthy();
});
 
//snapshot
it("matches snapshot on initial render", async () => {
  const tree = renderer.create(<EditProfileScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});
 //snapshot
it("matches snapshot after selecting refund policy", () => {
  const { getByText } = render(<EditProfileScreen />);
  fireEvent.press(getByText("REFUNDABLE"));
  const tree = renderer.create(<EditProfileScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});
 //snapshot
it("matches snapshot after filling form", () => {
  const { getByPlaceholderText, getByTestId } = render(<EditProfileScreen />);
  fireEvent.changeText(getByPlaceholderText("Enter your name"), "Ali");
  fireEvent.changeText(
    getByPlaceholderText("Enter phone number"),
    "3012345678"
  );
  fireEvent.changeText(getByTestId("input-address"), "New Address");
  const tree = renderer.create(<EditProfileScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("description mirrors address field input", () => {
  const { getByTestId } = render(<EditProfileScreen />);
  fireEvent.changeText(getByTestId("input-address"), "Shared Value");
  expect(getByTestId("input-description").props.value).toBe("Shared Value");
});

it("disables SAVE button if mandatory fields are empty", () => {
  (patchUpdateProfile as jest.Mock).mockClear();

  const { getByText } = render(<EditProfileScreen />);
  fireEvent.press(getByText("SAVE"));
  expect(patchUpdateProfile).not.toHaveBeenCalled();
});

it("does not call API on SAVE if no data is changed", async () => {
  (patchUpdateProfile as jest.Mock).mockClear();

  const { getByText } = render(<EditProfileScreen />);
  fireEvent.press(getByText("SAVE"));

  await waitFor(() => {
    expect(patchUpdateProfile).not.toHaveBeenCalled();
  });
});

it("loads initials correctly for lowercase names", async () => {
  (store.getSecureData as jest.Mock).mockResolvedValue(
    JSON.stringify({
      name: "ali",
      email: "ali@test.com",
      phoneNumber: "123",
      address: "",
      userId: "id",
    })
  );

  const { findByText } = render(<EditProfileScreen />);
  expect(await findByText("A")).toBeTruthy();
});

it("shows alert if save fails due to network error", async () => {
  (patchUpdateProfile as jest.Mock).mockRejectedValue(
    new Error("Network issue")
  );

  const { getByText, getByPlaceholderText, getByTestId } = render(
    <EditProfileScreen />
  );
  fireEvent.changeText(getByPlaceholderText("Enter your name"), "Ali");
  fireEvent.changeText(getByTestId("input-address"), "Somewhere");
  fireEvent.press(getByText("SAVE"));

  await waitFor(() => {
    expect(global.alert).toHaveBeenCalledWith(
      "Failed to save profile. Please try again."
    );
  });
});

it("renders the title 'Edit Profile' at the top", async () => {
  const { findByText } = render(<EditProfileScreen />);
  expect(await findByText("Edit Profile")).toBeTruthy();
});

it("displays all three refund policy options", () => {
  const { getByText } = render(<EditProfileScreen />);
  expect(getByText("REFUNDABLE")).toBeTruthy();
  expect(getByText("NON-REFUNDABLE")).toBeTruthy();
  expect(getByText("PARTIALLY REFUNDABLE")).toBeTruthy();
});

it("renders gender selection options", () => {
  const { getByText } = render(<EditProfileScreen />);
  expect(getByText("MALE")).toBeTruthy();
  expect(getByText("FEMALE")).toBeTruthy();
  expect(getByText("TRANSGENDER")).toBeTruthy();
});

it("shows country placeholder as 'Pakistan'", () => {
  const { getByPlaceholderText } = render(<EditProfileScreen />);
  const countryInput = getByPlaceholderText("Pakistan");
  expect(countryInput.props.editable).toBe(false); // ensure it's disabled
});

it("allows entering email address", () => {
  const { getByPlaceholderText } = render(<EditProfileScreen />);
  const emailInput = getByPlaceholderText("Enter your email");
  fireEvent.changeText(emailInput, "test@email.com");
  expect(emailInput.props.value).toBe("test@email.com");
});

});
