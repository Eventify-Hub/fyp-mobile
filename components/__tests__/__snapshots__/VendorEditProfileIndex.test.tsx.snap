// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditProfileScreen matches snapshot after filling form 1`] = `null`;

exports[`EditProfileScreen matches snapshot after selecting refund policy 1`] = `null`;

exports[`EditProfileScreen matches snapshot on initial render 1`] = `null`;
